### DISCUSSION TITLE
How to use nng with libev or libevent?

### DISCUSSION THREAD

**Discussion by mydream59** [nanomsg/nng#1833] [Q&A] [answered]
Encountered some issues when introducing nng, and would appreciate some help. Thank you very much.

A large number of ordinary sockets and libev are currently used in the original system. Now we want to gradually introduce nng, so it is necessary for the two to work together.

There are several points that need attention:
1. Need to support all platforms including Linux, Windows, Android, Mac, and iOS.
2. It will use TCP/UDP sockets and NNG's inter-process and inter-thread communication mechanisms (the focus is on cross-platform support and managing with libev).

The solution found is to obtain the fd of the nng socket and manage it with libev. However, since nng does not use sockets for inter-process or inter-thread communication, it is not possible to obtain the fd.

**Comment by gdamore** [comment, marked as answer]
NNG specifically has support for this.

See the NNG_OPT_SENDFD, and NNG_OPT_RECVFD options that you can use with a socket.

These will poll ready (poll for read) when the socket can either send or receive.  Note that these file descriptors are *only* for signaling and should not be accessed directly except for epoll() or poll() or select() signaling.

See https://nng.nanomsg.org/man/v1.2.2/nng_options.5.html#NNG_OPT_SENDFD and https://nng.nanomsg.org/man/v1.2.2/nng_options.5.html#NNG_OPT_RECVFD

**Reply by gdamore** [reply]
Note that integration with epoll in this way does incur overhead.  You may be stuck due to libev or something like that.

This should probably not be used on Windows unless you're using `WSASelect`.  Windows has better solutions, but I don't know if libev can work with them.  (WSASocket is really only meant to multiplex on sockets.
